### ISSUE TITLE
[1.8.0] New Envoy inbound cluster name format can lead to duplicated cluster issue

### ISSUE THREAD

**Issue by Laci21** [istio/istio#29199]
**Bug description**

I have a pod with two containers in it and using two different container ports. I have two services, both selecting the same pod, but the first service is targeting the first container and the second service is targeting the second container.

No matter which of those services I'm calling **always the same container will receive the request** in the pod.

[x] Docs
[x] Networking
[x] User Experience

**Expected behavior**

I would like to be able to route to both containers inside the pod, just like it was working before Istio 1.8.0.

**Steps to reproduce the bug**

1. Make sure that you install the following deployment (and services) to a namespace where istio injection is enabled:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo
  labels:
    k8s-app: echo
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      k8s-app: echo
  template:
    metadata:
      labels:
        k8s-app: echo
    spec:
      terminationGracePeriodSeconds: 2
      containers:
      - name: echo-backend
        image: k8s.gcr.io/echoserver:1.10
        ports:
        - containerPort: 8080
      - name: echo-frontend
        image: k8s.gcr.io/echoserver:1.10
        ports:
        - containerPort: 8081
---
apiVersion: v1
kind: Service
metadata:
  name: echo-backend
  labels:
    k8s-app: echo
  namespace: default
spec:
  ports:
  - name: http
    port: 80
    targetPort: 8080
  selector:
    k8s-app: echo
---
apiVersion: v1
kind: Service
metadata:
    name: echo-frontend
    labels:
        k8s-app: echo
    namespace: default
spec:
    ports:
    - name: http
      port: 80
      targetPort: 8081
    selector:
        k8s-app: echo
```

You should see in the `istiod` logs the following warning right away:
```
    "pilot_duplicate_envoy_clusters": {
        "inbound|80||": {
            "proxy": "echo-6dbb78ddc4-wr4bq.default",
            "message": "Duplicate cluster inbound|80|| found while pushing CDS"
        }
    }
```

1. Run a curl pod e.g. with: `kubectl run curl-test --image=radial/busyboxplus:curl -i --tty --rm`

1. Inside the curl pod, run: `curl echo-frontend.default.svc.cluster.local:80` and `curl echo-backend.default.svc.cluster.local:80`

1. If you check the logs of the echo pod's two app containers (`echo-frontend` and `echo.backend`) you should see that both calls went to the same container (instead of one request to each)

**Additional info**

If you check the config dump with something like `kubectl port-forward echo-6dbb78ddc4-wr4bq 15000`, you will only see this cluster:
```
      "load_assignment": {
       "cluster_name": "inbound|80||",
       "endpoints": [
        {
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "127.0.0.1",
              "port_value": 8080
             }
            }
           }
          }
         ]
        }
```

In older versions the config dump looked like this and this bug was not present:
```
      "load_assignment": {
       "cluster_name": "inbound|80|http|echo-backend.default.svc.cluster.local",
       "endpoints": [
        {
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "127.0.0.1",
              "port_value": 8080
             }
            }
           }
          }
         ]
        }
...
      "load_assignment": {
       "cluster_name": "inbound|80|http|echo-frontend.default.svc.cluster.local",
       "endpoints": [
        {
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "127.0.0.1",
              "port_value": 8081
             }
            }
           }
          }
         ]
        }
```

So to us it seems like this deliberate [inbound cluster name format change](https://istio.io/latest/news/releases/1.8.x/announcing-1.8/upgrade-notes/#inbound-cluster-name-format) is causing issues in this case.
Related PR: https://github.com/istio/istio/pull/27303. cc @howardjohn 

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**

```
$ istioctl version --remote
client version: 1.7.3
control plane version: 1.8.0
data plane version: 1.8.0 (33 proxies)
```

```
$ kubectl version --short
Client Version: v1.18.5
Server Version: v1.18.6
```

**How was Istio installed?**

Installed by [Backyards](https://banzaicloud.com/products/backyards/), which installs Istio using the [Banzai Cloud Istio operator](https://github.com/banzaicloud/istio-operator). I did only one tweak regarding the installation, which is that I changed the Docker image for `istiod` to the upstream image to recreate this issue: `docker.io/istio/pilot:1.8.0`. The issue should happen with any other Istio 1.8.0 installation.

**Environment where bug was observed (cloud vendor, OS, etc)**

K8s cluster: Amazon [PKE](https://github.com/banzaicloud/pke) 1.18.6
OS image: `Ubuntu 18.04.4 LTS`
Kernel version: `4.15.0-1065-aws`
Container runtime: `containerd://1.3.3`

**Comment by khernyo** [comment]
I just tested it on a pure istio install, same result.

One thing I noticed is that the requests are randomly distributed between the two containers. So, for
example `while true; do curl echo-backend.default.svc.cluster.local:80; done` will hit both containers,
which is also unexpected. These requests should only end up in the "echo-backend" container.

Setup and versions:

```
$ kind version
kind v0.9.0 go1.15.2 linux/amd64

$ kind create cluster

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:17:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}

$ istioctl install

$ istioctl version
client version: 1.8.0
control plane version: 1.8.0
data plane version: 1.8.0 (1 proxies)

$ kubectl apply -f resources.yaml
```

`resources.yaml` contains the resources from the above comment by @Laci21

**Comment by howardjohn** [comment]
cc @istio/release-managers-1-8

**Comment by howardjohn** [comment]
@Laci21 thanks for the report. I have confirmed everything you said and captured it in a regression test in https://github.com/istio/istio/pull/29201. I am disappointed we didn't catch this earlier..

I think the solution here is the cluster port should be `inbound|workload-port||`, rather than `inbound|service-port||`. This is consistent with the inbound listener configuration.

So the flow would be:
```go
instancePorts := unique(port.targetPort for port in servicesForProxy)
for port in instancePorts {
  add cluster for `inbound|port` 
  // If there are multiple service ports mapping to this cluster, we put both Service attributes in the metadata
  // If there are different protocols defined on the multiple services, we pick one. This has always been invalid config
  // If multiple services with DR, we pick an arbitrary one. This matches today, and we are fixing in other work

  add listener config for `dest_port=port`, forward to `inbound|port`
}
```

Basically now this makes the inbound filter chains map 1:1 with the cluster, whereas today we may have a single cluster for multiple filter chains, or the case in this issue where its broken.

@bianpengyuan @yangminzhu @mandarjog does this make sense? I can't think of any cases where it would break down but it seems "too easy" so I am a bit suspicious I am missing something... I added a naive implementation in https://github.com/istio/istio/pull/29203 so we will see how that does

**Comment by bianpengyuan** [comment]
How does the cluster telemetry metadata look like after switching to workload port? Want to make sure the correct service info is attached to each cluster.

**Comment by howardjohn** [comment]
@bianpengyuan I think its mostly the same as today.

Today, we take the cluster (service) port and lookup all Services matching, and add them.

In the change suggested above, we would take the cluster (workload) port and lookup all Services matching, and add them.

Note my PR above doesn't actually do that yet, I just wanted to quickly see if all the networking and security tests pass first

**Comment by bianpengyuan** [comment]
> In the change suggested above, we would take the cluster (workload) port and lookup all Services matching, and add them.

Yeah, this sounds sense to me as long as we make sure that only services target that workload port are attached in the metadata. This make sure telemetry does not have regression from 1.7.

**Comment by yangminzhu** [comment]
> @bianpengyuan I think its mostly the same as today.
> 
> Today, we take the cluster (service) port and lookup all Services matching, and add them.
> 
> In the change suggested above, we would take the cluster (workload) port and lookup all Services matching, and add them.
> 
> Note my PR above doesn't actually do that yet, I just wanted to quickly see if all the networking and security tests pass first

So the change is only in the cluster? Does the change affect the `portLevelSettings` in the DestinationRule when configuring the TLS setting in the cluster in client side?

**Comment by howardjohn** [comment]
> portLevelSettings

This applies to outbound cluster only. The changes mentioned above are strictly about inbound cluster.

**Comment by yangminzhu** [comment]
> > portLevelSettings
> 
> This applies to outbound cluster only. The changes mentioned above are strictly about inbound cluster.

Ah, you're right, then I think it shouldn't affect the security, we don't do anything on the inbound cluster.

**Comment by howardjohn** [comment]
Ok makes sense. Thanks for verifying

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/29203 should fix this an is passing tests.

**Comment by hzxuzhonghu** [comment]
How about multi services with the same target port? How can the metrics be distinguished?

**Comment by howardjohn** [comment]
@hzxuzhonghu this is already an issue on all versions, we cannot know for sure the service

**Comment by hzxuzhonghu** [comment]
IC, even for HTTP it is impossible to distinguish currently

**Comment by Laci21** [comment]
Thank you @howardjohn for the quick fix!
